### PR TITLE
[JENKINS-26755] cache node environment

### PR DESCRIPTION
**Pull request by ndeloof**
[JENKINS-26755](https://issues.jenkins-ci.org/browse/JENKINS-26755)

prevent inefficient channel overuse

### CODE REVIEW

**Comment by ndeloof** [comment]
@reviewbybees also see RM-2851

File: core/src/main/java/hudson/model/Computer.java [hunk 1]
```diff
@@ -162,11 +162,17 @@
...
+     * @see #getEnvironment()
+     */
+    private transient EnvVars cachedEnvironment;
+
     private volatile boolean hostNameCached;
 
     private final WorkspaceList workspaceList = new WorkspaceList();
 
-    protected transient List<Action> transientActions;
+    protected volatile List<Action> transientActions;
```
Comment (stephenc) [inline core/src/main/java/hudson/model/Computer.java]: seriously? this is no longer transient?

File: core/src/main/java/hudson/model/Computer.java [hunk 2]
```diff
@@ -939,7 +946,12 @@ public String getSearchUrl() {
      * If this is the master, it returns the system property of the master computer.
      */
     public EnvVars getEnvironment() throws IOException, InterruptedException {
-        return EnvVars.getRemote(getChannel());
+        if (cachedEnvironment != null) {
+            return cachedEnvironment;
```
Comment (stephenc) [inline core/src/main/java/hudson/model/Computer.java]: what if there was a write setting this to null inbetween the check of null

File: core/src/main/java/hudson/model/Computer.java [hunk 3]
```diff
@@ -939,7 +946,12 @@ public String getSearchUrl() {
...
+        }
+
+        cachedEnvironment = EnvVars.getRemote(getChannel());
```
Comment (stephenc) [inline core/src/main/java/hudson/model/Computer.java]: Again, you should be returning the value you just built and not the value from the volatile

**Comment by stephenc** [comment]
@ndeloof you need to re-read the [train book](http://books.google.ie/books/about/Java_Concurrency_in_Practice.html) ;-)

File: core/src/main/java/hudson/model/Computer.java (hunk 1 above)
Comment (ndeloof) [inline core/src/main/java/hudson/model/Computer.java]: :-\

File: core/src/main/java/hudson/model/Computer.java (hunk 2 above)
Comment (ndeloof) [inline core/src/main/java/hudson/model/Computer.java]: indeed, need to rely on a second volatile "environmentCached" flag (like cachedHostName does).

**Comment by ndeloof** [comment]
@stephenc I even contributed to write it's french translation, but Java threading is such a complex stuff (at least for me)...

File: core/src/main/java/hudson/model/Computer.java (hunk 2 above)
Comment (stephenc) [inline core/src/main/java/hudson/model/Computer.java]: Nope. Just store the value in a local field and return that if non-null.

The cached value should be much like String's hashCode field, in that, as long as non-null it should be equal to a re-run.

Comment (stephenc) [inline core/src/main/java/hudson/model/Computer.java]: If you are using a second field (i.e. a cachedEnv) then you really want instead to use a lock and avoid volatile at all. You cannot safely manage atomically updating the state of two volatiles at the same time. Where state is to be updated atomically you either use a volatile reference to an immutable state object or you use a lock to guard the state fields

Comment (stephenc) [inline core/src/main/java/hudson/model/Computer.java]: Something like this:

```
 public EnvVars getEnvironment() throws IOException, InterruptedException {
        EnvVars cachedEnvironment = this.cachedEnvironment; 
        if (cachedEnvironment != null) {
            return cachedEnvironment;
        cachedEnvironment = EnvVars.getRemote(getChannel());
        // all non-null values should be equivalent so in a race it doesn't matter who wins
        this.cachedEnvironment = cachedEnvironment;
        return cachedEnvironment;
    }
```

File: core/src/main/java/hudson/model/Computer.java [hunk 4]
```diff
@@ -164,6 +164,13 @@
     private volatile String cachedHostName;
     private volatile boolean hostNameCached;
 
+    /**
+     * @see #getEnvironment()
+     */
+    private volatile EnvVars cachedEnvironment;
+    private volatile boolean environmentCached;
```
Comment (stephenc) [inline core/src/main/java/hudson/model/Computer.java]: Do not use two volatiles to hold state that needs to be atomically updated

:-1:

Comment (ndeloof) [inline core/src/main/java/hudson/model/Computer.java]: @stephenc so you mean `cachedHostName` is wrong as well. Any suggestion for a better pattern ?

Comment (stephenc) [inline core/src/main/java/hudson/model/Computer.java]: Yes it is also an incorrect pattern. The question with cachedHostName is whether null implies that the host name was cached or whether it was tried and failed. 

If you have dual meaning from the `null` return value then you need to have a separate object state or just use a fricking lock.

**Comment by jglick** [comment]
:+1:

**Comment by jglick** [comment]
You forgot to update `changelog.html` and close the JIRA issue.

Also BTW by @reviewbybees rules you were missing a :+1: to merge.

**Comment by stephenc** [comment]
here's the :+1: he was missing

(if you are not a cloudbees employee, you don't need to pay any attention to the `@reviewbybees` rules they are for cloudbees employees only ;-) )

**Comment by daniel-beck** [comment]
This may have caused [JENKINS-27178](https://issues.jenkins-ci.org/browse/JENKINS-27178).

**Comment by daniel-beck** [comment]
And [JENKINS-27188](https://issues.jenkins-ci.org/browse/JENKINS-27188).

**Comment by daniel-beck** [comment]
And [JENKINS-27199](https://issues.jenkins-ci.org/browse/JENKINS-27199).

**Comment by daniel-beck** [comment]
And [JENKINS-27739](https://issues.jenkins-ci.org/browse/JENKINS-27739).

**Pull request merged**